File: Behavioral/Lab20/C#/Entities/States/RentableState.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Interfaces;

namespace C_.Entities.States
{
    public class RentableState : EstateState
    {
        public RentableState(RealEstate estate) : base(estate){}
        public override void Rent(Customer customer, DateTime startDate, DateTime endDate)
        {
            if(!_estate.IsAlreadyRented){
                Console.WriteLine($"Customer {customer.NameAndSurname} rented an estate from {startDate} to {endDate}");
                _estate.IsAlreadyRented = true;    
            }
            else
                Console.WriteLine($"Customer {customer.NameAndSurname} can't rent an estate because it is already rented");
        }

        public override void Sell(Customer customer)
        {
            Console.WriteLine($"Customer {customer.NameAndSurname} can't sell an estate");
            
        }

        public override void SetEstateToBuy(Customer customer, double price)
        {
            Console.WriteLine($"Customer {customer.NameAndSurname} can't buy an estate");
        }

    }
}

File: Behavioral/Lab20/Typescript/src/Entities/States/rentable_state.ts
import { Customer } from "../customer";
import { RealEstate } from "../real_estate";
import { EstateState } from "./estate_state";

export class RentableState extends EstateState{
    public rent(customer: Customer, startDate: Date, endDate: Date): void {
        if(!this._estate.isAlreadyRented){
            console.log(`Customer ${customer.nameAndSurname} rented an estate from ${startDate} to ${endDate}`);
            this._estate.isAlreadyRented = true;    
        }
        else
            console.log(`Customer ${customer.nameAndSurname} can't rent an estate because it is already rented`);
    }

    public sell(customer: Customer): void {
        console.log(`Customer ${customer.nameAndSurname} can't sell an estate`);
    }

    public setEstateToBuy(customer: Customer, price: number): void {
        console.log(`Customer ${customer.nameAndSurname} can't buy an estate`);
    }
}

